R/France.R

```diff
@@ -0,0 +1,142 @@
+#' France Class containing country specific attributes and methods
+#'
+#' @description Country specific information for downloading, cleaning
+#' and processing covid-19 region data for France.
+#'
+#' Data is sourced from \url{https://health-infobase.France.ca}.
+#' @details Inherits from `DataClass`
+#' @examples
+#' \dontrun{
+#' region <- France$new(verbose = TRUE, steps = TRUE)
+#' region$download()
+#' region$clean()
+#' region$process()
+#' region$return()
+#' }
+France <- R6::R6Class("France",
+  inherit = DataClass,
+  public = list(
+
+    # Core Attributes (amend each paramater for country specific infomation)
+    #' @field level_1_region the level 1 region name.
+    level_1_region = "region",
+    #' @field level_2_region the level 1 region name.
+    level_2_region = "departement",
+    #' @field data_url link to raw data for cases
+    data_url = "https://www.data.gouv.fr/fr/datasets/r/406c6a23-e283-4300-9484-54e78c8ae675", # nolint
+    #' @field source_data_cols existing columns within the raw data
+    source_data_cols = c("cases_new", "deaths_new",
+                         "hosp_new", "tested_new"),
+    #' @field hosp_url link to raw data for hospitalisations
+    hosp_url = "https://www.data.gouv.fr/fr/datasets/r/6fadff46-9efd-4c53-942a-54aca783c30c", # nolint
+
+
+    #' @description France-specific function for downloading raw data.
+    download = function() {
+      if (self$level == 1) {
+        self$data_url <- "https://www.data.gouv.fr/fr/datasets/r/001aca18-df6a-45c8-89e6-f82d689e6c01" # nolint
```

Comment:
can we add this link as a new field?

Reply:
Do we want to? Or do we want to have `data_url` which we always use (for both levels), but which points to different things depending on level, and `hosp_url` which we only use for one level. I had gone with this approach, instead of explicitly labelling `data_url_level_1` and `data_url_level_2`; if we did that, then we'd no longer have `data_url`.

Reply:
Yes that is a good point. My comment was mainly driven by not wanting links in the body of the code but I see your reasoning. Perhaps we have a new field for the level 2 data (or one for both) and then overwrite `data_url` in `download` depending on what type of data is requested? What do you think @joseph-palmer?

Reply:
Something additional which could be done here - and I don't know if this is just a France thing or if it's seen elsewhere. Currently the list of fields available from original differs for level 1 and level 2, with more detail available at level 2. It may be possible to set up aggregating code like in the Lithuania model which could let us provide this data at level 1 as well.

But this still leaves open how we describe which fields are available in the source, and how the tests look at this (if they do).

Reply:
Currently, tests go through each class and check for a `data_url` field so as long as we have this the class structure tests will be happy. I think all urls should be stored as public fields so we could do this a few different ways:

1. have 2 fields: `data_url` and `data_level2_url` where the level 1 url sits in `data_url` and level 2 sits in `data_level2_url`, this could then be explained in the description and just direct which one to use if level is 1 or 2.  
2. Alternatively, `data_url` just has to exist as a field so rather than making this a string could just have it as a named list: `data_url = list(1 = "path/to/level/1/data", 2 = "path/to/level/2/data")` then we could just pass `self$level` into `data_url` like `data_url[[self$level]]`. 

Both have their merits, I'm more drawn to option 2 as I thinks its cleaner, but happy with whatever works, just as long as links are not stored in method code.

Reply:
Option to is nice and makes lots of sense if `data-raw` becomes a list.

Reply:
So, sorry, I missed this and wrote something lengthy into #212

I think it's useful to be able to avoid unnecessary downloads - if there are things which we don't need to fetch to do level 2, let's not fetch them. For testing purposes we'll need to download abbreviated versions of the data for both levels, store them separately, then mock them into `self$data$raw` differently.

It's also going to be very useful to class authors to be able to predict/determine how downloaded data is going to be labelled, so we can usefully have something like `self$data$raw`, `self$data$raw_hosp` and `self$data$raw_tests`; or `self$data$raw$cases`, `self$data$raw$hosp` and `self$data$raw$tests`. (Perhaps the default should be `self$data$raw$all` for that rare case where we have a country which will provide deaths data but not cases data at a particular regional level.)

Reply:
For now I have split things out so we have `data_url` (for level 1 data) and `level_2_data_url` (for cases) and `level_2_hosp_url` (for hospitalization).